common/src/main/java/com/gentics/mesh/graphdb/spi/Database.java

```diff
@@ -160,42 +161,68 @@ default Completable asyncTx(TxAction0 txHandler) {
 	}
 
 	/**
-	 * Executes a supplier in a transaction within the worker thread pool.
-	 * If the supplier returns null, the maybe is completed, else the value is returned.
-	 * @param handler
+	 * Executes a supplier in a transaction within the worker thread pool. If the supplier returns null, the maybe is completed, else the value is returned.
+	 * 
 	 * @param <T>
+	 * @param handler
 	 * @return
 	 */
 	default <T> Maybe<T> maybeTx(Function<Tx, T> handler) {
+		return maybeTx(handler, false);
+	}
+
+	/**
+	 * Executes a supplier in a transaction within the worker thread pool. If the supplier returns null, the maybe is completed, else the value is returned.
+	 * 
+	 * @param handler
+	 * @param useWriteLock
+	 *            Whether to apply a write lock around the transaction
+	 * @param <T>
+	 * @return
+	 */
+	default <T> Maybe<T> maybeTx(Function<Tx, T> handler, boolean useWriteLock) {
 		return new io.vertx.reactivex.core.Vertx(vertx()).rxExecuteBlocking(promise -> {
 			try {
-				promise.complete(tx(handler::apply));
+				if (useWriteLock) {
+					T result = null;
+					try (WriteLock lock = writeLock().lock(null)) {
+						result = tx(handler::apply);
+					}
+					promise.complete(result);
+				} else {
+					promise.complete(tx(handler::apply));
+				}
+
 			} catch (Throwable e) {
 				promise.fail(e);
 			}
 		}, false);
 	}
 
 	/**
-	 * Executes a supplier in a transaction within the worker thread pool.
-	 * If the supplier returns null, a {@link java.util.NoSuchElementException} is emitted.
+	 * Executes a supplier in a transaction within the worker thread pool. If the supplier returns null, a {@link java.util.NoSuchElementException} is emitted.
+	 * 
 	 * @param handler
 	 * @param <T>
 	 * @return
 	 */
 	default <T> Single<T> singleTx(Supplier<T> handler) {
-		return maybeTx(tx -> handler.get()).toSingle();
+		return maybeTx(tx -> handler.get(), false).toSingle();
+	}
+
+	default <T> Single<T> singleTxWriteLock(Function<Tx, T> handler) {
+		return maybeTx(handler, true).toSingle();
 	}
 
 	/**
-	 * Executes a supplier in a transaction within the worker thread pool.
-	 * If the supplier returns null, a {@link java.util.NoSuchElementException} is emitted.
+	 * Executes a supplier in a transaction within the worker thread pool. If the supplier returns null, a {@link java.util.NoSuchElementException} is emitted.
+	 * 
 	 * @param handler
 	 * @param <T>
 	 * @return
 	 */
 	default <T> Single<T> singleTx(Function<Tx, T> handler) {
-		return maybeTx(handler).toSingle();
+		return maybeTx(handler, true).toSingle();
```

Comment:
Is this intended? The above `singleTx` does not use write locks by default

Reply:
WTF. No. Thanks!